_posts/2016-05-11-android-testing-with-kotlin.md

```diff
@@ -0,0 +1,422 @@
+compiler aware of them.
+
+```java
+@Test
+fun myTestCase() { ... }
+```
+
+## Espresso tests
+
+Making use of the interoperability of Java and Kotlin code we can blend Espresso into
+our Kotlin test classes in the blink of an eye. That's how we can benefit both from
+battle tested Espresso framework that makes testing UI components a bliss and
+all the gems that Kotlin provides as a language to make writing our _glue_ code and other
+testing logic as easy and clean as possible.
+
+As usual, we store our Espresso and other UI tests under `src/androidTest`, with a
+little discrepancy that instead of `java` subdirectory, we store our Kotlin source code in
+`kotlin` subdirectory. And in order to run Espresso tests we need to declare it as a dependency
+in our build script plus set `AndroidJUnitRunner` as the default test instrumentation
+runner, exactly like you would do with plain Espresso tests:
+
+```groovy
+android {
+    ...
+    defaultConfig {
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+    }
+}
+
+dependencies {
+    ...
+    androidTestCompile 'com.android.support.test.espresso:espresso-core:2.2.2'
+}
+```
+
+Now to the test suite itself. Our example Espresso test class
+[`EspressoKotlinTest`](https://github.com/GreenhouseCI/kotlin-tests/blob/master/app/src/androidTest/kotlin/com/greenhouseci/kotlin_tests/kotlintests/EspressoKotlinTest.kt)
+looks like
+
+```ruby
+@RunWith(AndroidJUnit4::class)
+class EspressoKotlinTest {
+
+    ...
+
+    @Rule @JvmField
+    var activityRule = ActivityTestRule<MainActivity>(MainActivity::class.java)
+
+    ...
+
+    @Test
+    fun testAddItemsToList() {
+        val formatter = Formatter()
+        val lv = onData(anything()).inAdapterView(withId(R.id.lv_items))
+
+        var items = arrayOf(
+            Pair("    item 1", "item 1"),
+            Pair("item 2  ", "item 2"),
+            Pair("  item  \n3  ", "item 3")
+        )
+
+        items.map { pair -> addItemToList(pair.first) }
+        items.mapIndexed { index, pair ->
+            val expected = formatter.strip(pair.second)
+            lv.atPosition(index).onChildView(withId(R.id.tv_list_item)).check(
+                matches(withText(expected))
+            )
+        }
+    }
+}
+```
+
+You can see that the Espresso test suite at hand doesn't differ much from the
+unit tests we discussed in the previous section. And neither does it differ much from
+your already existing Espresso tests written in Java. Truth is, that majority of the
+code that does the heavy lifting is plain *Java* and other parts are almost
+drop in replacements.
+
+As usual with instrumentation tests, we start off by creating an instrumented JUnit 4
+test class by adding
+
+```ruby
+@RunWith(AndroidJUnit4::class)
+```
+
+annotation at the beginning of our test class definition. And naturally we can make
+use of Espresso's built-in `ActivityTestRule` to handle our activity life cycle:
```

Comment:
Generally, avoid starting sentences with "And" or "But". Here, can just drop the "And" - "Naturally, we can..."